File: frontend/src/types/strategy.ts
export interface AssetAllocation {
  symbol: string;
  name: string;
  weight: number; // e.g., 0.20 for 20%
  contractAddress: `0x${string}`;
  verified: boolean;
}

export interface StrategyRule {
  condition_type: 'PRICE_DROP' | 'PERCENT_GAIN' | 'POSITION_ABOVE';
  target_symbol?: string;
  threshold: number;
  action_type: 'BUY' | 'SELL' | 'REBALANCE';
  action_value?: number;
}

export interface StrategyPolicy {
  name: string;
  capital_usdc: number;
  assets: AssetAllocation[];
  reserve: number; // e.g. 0.10 for 10%
  rebalance_days: number;
  rules: StrategyRule[];
}

export type PipelineStep = 'INPUT' | 'COMPILING' | 'COMPILED' | 'PREVIEW' | 'EXECUTING' | 'SUCCESS';

File: src/domain/strategy.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StrategyPolicy {
    pub name: String,
    pub capital_usdc: f64,
    pub assets: Vec<AssetAllocation>,
    pub reserve: f64, // e.g., 0.10 for 10%
    pub rebalance_days: u32,
    pub rules: Vec<StrategyRule>,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct AssetAllocation {
    pub symbol: String,
    pub weight: f64, // e.g., 0.20 for 20%
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct StrategyRule {
    pub condition_type: String, // "PRICE_DROP", "POSITION_ABOVE"
    pub target_symbol: Option<String>,
    pub threshold: f64,
    pub action_type: String, // "BUY", "SELL", "REBALANCE"
    pub action_value: Option<f64>, 
}